lighthouse-core/lib/network-request.js

```diff
@@ -0,0 +1,244 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+ * @fileoverview Fills most of the role of NetworkManager and NetworkRequest classes from DevTools.
+ * @see https://cs.chromium.org/chromium/src/third_party/blink/renderer/devtools/front_end/sdk/NetworkRequest.js
+ * @see https://cs.chromium.org/chromium/src/third_party/blink/renderer/devtools/front_end/sdk/NetworkManager.js
+ */
+
+const URL = require('./url-shim');
+const resourceTypes = require('../../third-party/devtools/ResourceType').TYPES;
+
+module.exports = class NetworkRequest {
+  constructor() {
+    this.requestId = '';
+    this._requestId = '';
+    this.connectionId = '';
+    this.connectionReused = false;
+
+    this.url = '';
+    this._url = '';
+    this.protocol = '';
+    this.parsedURL = /** @type {LH.WebInspector.ParsedURL} */ ({scheme: ''});
+
+    this.startTime = -1;
+    /** @type {number} */
+    this.endTime = -1;
+    /** @type {number} */
+    this._responseReceivedTime = -1;
+
+    this.transferSize = 0;
+    this._transferSize = 0;
+    this._resourceSize = 0;
+    this._fromDiskCache = false;
+    this._fromMemoryCache = false;
+
+    this.finished = false;
+    this.requestMethod = '';
+    this.statusCode = -1;
+    this.redirectSource = /** @type {NetworkRequest|undefined} */ (undefined);
+    this.redirectDestination = /** @type {NetworkRequest|undefined} */ (undefined);
+    this.redirects = /** @type {NetworkRequest[]|undefined} */ (undefined);
+    this.failed = false;
+    this.localizedFailDescription = '';
+
+    this._initiator = /** @type {LH.Crdp.Network.Initiator} */ ({type: 'other'});
+    this._timing = /** @type {LH.Crdp.Network.ResourceTiming|undefined} */ (undefined);
+    this._resourceType = /** @type {LH.WebInspector.ResourceType|undefined} */ (undefined);
+    this._mimeType = '';
+    this.priority = () => /** @type {LH.Crdp.Network.ResourcePriority} */ ('Low');
+    this.initiatorRequest = () => /** @type {NetworkRequest|undefined} */ (undefined);
+    this._responseHeaders = /** @type {LH.WebInspector.HeaderValue[]} */ ([]);
+
+    this._fetchedViaServiceWorker = false;
+    this._frameId = /** @type {string|undefined} */ ('');
+    this._isLinkPreload = false;
+
+    // Make sure we're compitable with NetworkRequest
```

Comment:
compatible

Reply:
lol whoops, at least it was one of the more amusing ones :)